src/codec/dogstatsd.rs

```diff
@@ -0,0 +1,795 @@
+                break;
+            }
+            Some(_) => (),
+            None => return Err(invalid()),
+        }
+    }
+
+    // Value(s) - <VALUE1>:<VALUE2>
+    let mut values = Vec::new();
+    loop {
+        match d.next() {
+            Some((idx, b':' | b'|')) => {
+                let s = substr(data, section_start..idx)?;
+                let v: f64 = s.parse()?;
+                let value = Value::from(v);
+                values.push(value);
+                section_start = idx + 1;
+
+                if substr(data, idx..=idx)?.eq("|") {
+                    break;
+                }
+            }
+            Some(_) => (),
+            None => return Err(invalid()),
+        }
+    }
+    m.insert("values".into(), Value::from(values));
+
+    // <TYPE>
+    match d.next() {
+        Some((i, b'c' | b'd' | b'g' | b'h' | b's')) => {
+            section_start = i + 1;
+            m.insert("type".into(), substr(data, i..=i)?.into());
+        }
+        Some((i, b'm')) => {
+            if let Some((j, b's')) = d.next() {
+                m.insert("type".into(), substr(data, i..=j)?.into());
+                section_start = i + 1;
+            } else {
+                return Err(invalid());
+            }
+        }
+        _ => return Err(invalid()),
+    };
+
+    // Optional Sections
+    let sections: Vec<&str> = substr(data, section_start..)?.split("|").collect();
+
+    for section in sections.iter() {
+        if section.starts_with('@') {
+            let sample_rate = &section[1..];
+            let sample_rate_float: f64 = sample_rate.parse()?;
+            m.insert("sample_rate".into(), Value::from(sample_rate_float));
+        } else if section.starts_with('#') {
+            let tags: Vec<&str> = section[1..].split(",").collect();
+            m.insert("tags".into(), Value::from(tags));
+        } else if section.starts_with('c') {
+            let container_id = &section[2..];
+            m.insert("container_id".into(), Value::from(container_id));
+        }
+    }
+
+    Ok(Value::from(m))
+}
+
+fn decode_event(data: &[u8]) -> Result<Value> {
+    let mut d = data.iter().enumerate();
+    let mut m = Object::with_capacity(13);
+    m.insert("dogstatsd_type".into(), Value::from("event"));
+    let section_start: usize;
+    let mut optional_sections = false;
+    let mut optional_text_idx = 0;
+
+    // Title/Text Lengths and Title
+    loop {
+        match d.next() {
+            Some((idx, b'|')) => {
+                let v: Vec<&str> = substr(data, 2..idx)?.split(":").collect();
+                let lens = v[0];
+                let len_vec: Vec<&str> = lens.split(",").collect();
+                let title_len: i32 = len_vec[0][1..].parse().unwrap();
+                let text_len: i32 = len_vec[1][0..len_vec[1].len() - 1].parse().unwrap();
+                let title = v[1];
+                m.insert("title_length".into(), Value::from(title_len));
```

Comment:
i'd say those lebngths don't need to be part of the decoded event, but are implicitly stored in the title and text values.

Reply:
Sounds good. At first, I didn't know how much we wanted to imply vs just take from the payload. This makes sense though so, I'll remove and calculate from the actual title/text